### PR TITLE
[orc8r][agw] Updating metrics collection gRPC message size limit / making it configurable

### PR DESCRIPTION
**Pull request by ardzoht**
## Summary

- Adding max-send-message-size metrics_collector client gRPC channel parameter
- Adding grpc_max_size to magmad.yml configs to configure value (default to send is 4MB)
- Updating metricsd gRPC server to allow receiving messages up to 50MB (default to receive is 4MB)

## Test Plan

- Tested on mcp02, with cwf02 and feg setup
- Tested locally with local orc8r and agw, setting configurable value 

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: cwf/gateway/configs/magmad.yml [hunk 1]
```diff
@@ -79,6 +79,7 @@ metricsd:
   sync_interval: 60 # How frequently to sync to cloud in seconds
   grpc_timeout: 30 # Timeout in seconds
   queue_length: 1000 # Number of failed samples to enqueue for resend
+  grpc_msg_size: 4 # Max message size for gRPC channel in MBs
```
Comment (xjtian) [inline cwf/gateway/configs/magmad.yml]: be explicit: `max_grpc_msg_size_mb`

File: orc8r/gateway/python/magma/magmad/metrics_collector.py [hunk 2]
```diff
@@ -34,8 +34,8 @@ class MetricsCollector(object):
     _services = []
 
     def __init__(self, services, collect_interval,
-                 sync_interval, grpc_timeout, queue_length, loop=None,
-                 post_processing_fn=None):
+                 sync_interval, grpc_timeout, grpc_msg_size,
+                 queue_length, loop=None, post_processing_fn=None):
```
Comment (xjtian) [inline orc8r/gateway/python/magma/magmad/metrics_collector.py]: type annotate these params while you're touching this constructor

File: orc8r/gateway/python/magma/magmad/metrics_collector.py [hunk 3]
```diff
@@ -183,6 +186,18 @@ def _get_uptime_metric(service_name, start_time):
     return family_proto
 
 
+def _get_metrics_chan_grpc_options(msg_size: int):
```
Comment (xjtian) [inline orc8r/gateway/python/magma/magmad/metrics_collector.py]: `msg_size_mb`

File: orc8r/gateway/python/magma/magmad/metrics_collector.py (hunk 2 above)
Comment (xjtian) [inline orc8r/gateway/python/magma/magmad/metrics_collector.py]: `grpc_max_msg_size_mb`

**Review by xjtian — CHANGES_REQUESTED** [review]
- Why set the client side payload limit to 4MB but the server side to 50?
- What is the behavior of the grpc client when the request is too large? Will we just drop that request on the client side? If the AGW is collecting a ton of metrics and every request is oversized, could we end up in a situation where we don't sync anything to metricsd?
- I think you need to combo this change with an update to the nginx container and Helm deployment yaml as well. http://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size might be the right param to tweak here, @mpgermano might have some older nginx logs you could look at and see what the exact error is when we send a massive request.

**Comment by ardzoht** [comment]
@xjtian 

- I wanted to define a maximum receiving message size on the server side, and then having the AGW define a configurable maximum message size to send, previously they were on default values (4MB each)
- If the request is oversized, then the collected samples will not be synced to metricsd, we've encountered this issue on mcp setup where with scalability testing due to the gRPC message exceeding the maximum allowed size, ideally I'm thinking on implementing doing batching for sending the metrics but for now to unblock this we want to increase it as a short-term fix.
- That makes sense, will look for the changes to apply for nginx configuration.

**Comment by xjtian** [comment]
Let's bump up the default send limit on the client side to something like 8MB in that case to give us some more headroom for the specific situations where we are experiencing oversized requests.

**Comment by ardzoht** [comment]
@xjtian Added client_max_size to nginx config to resolve 413 error response code.

**Review by xjtian — APPROVED** [review]

**Review by themarwhal — APPROVED** [review]

**Pull request merged**